dask/blockwise.py

```diff
@@ -655,21 +657,25 @@ def _inject_io_tasks(dsk, io_deps, output_indices, new_axes):
     # Loop through graph, and replace IO-placeholder tasks
     # with the actual underlying IO function
     for k in dsk:
-        for io_name, io_subgraph in io_deps.items():
-            # Leave out `new_axes` in key
-            io_key = (io_name,) + tuple(
-                [
-                    k[i + 1]
-                    for i, idx in enumerate(output_indices)
-                    if idx not in new_axes
-                ]
-            )
-            if io_key in dsk[k]:
-                # Inject IO-function arguments into the blockwise graph
-                # as a single (packed) tuple.
+        for i in range(1, len(dsk[k])):
+            io_key = dsk[k][i]
+            if isinstance(io_key, tuple) and len(io_key) and io_key[0] in io_deps:
+                io_subgraph = io_deps.get(io_key[0])
                 io_item = io_subgraph.get(io_key)
                 io_item = list(io_item[1:]) if len(io_item) > 1 else []
-                new_task = [io_item if v == io_key else v for v in dsk[k]]
+                new_task = []
+                for v in dsk[k]:
+                    # Add additional type check to short circuit equality check.
+                    # In some cases v can be array-like and want elementwise
+                    # comparisons over strict object equality, which is not what
+                    # we want here. For more discussion, see
+                    # https://github.com/dask/dask/pull/4805#discussion_r284312315
+                    if type(v) == type(io_key) and v == io_key:
```

Comment:
So, this is just a check that v (like `io_key`) is a tuple - correct?

Reply:
This is more meant as a check that `v` is not a numpy array (which winds up being the case in the test suite in a few places?)

Reply:
(although I meant to bring this back to a list comprehension after debugging, hang on)

Reply:
A minimal example which feeds a small numpy array into the task graph, and raises a spurious `FutureWarning` without this check:
```python
from operator import add
import dask.array as da
import numpy as np

x = da.ones((10, 10), chunks=(5, 5))
y = np.ones((5, 5))

z = x.map_blocks(add, y, dtype=x.dtype)
z.compute()
```

Happy to consider other workarounds

Reply:
Okay cool - I see.  I'm not sure of a better way to do this.  I guess, since we already know the index (`i`) of the key in `dsk[k]`, we could avoid any type checking and/or comparison, and just replace a specific index?

Reply:
Wouldn't we still need to test somewhere whether a given task should be replaced?

Reply:
I could be confused, but I think we already know that the "ith" element in `dsk[k]` needs to be replaced once we get the to `new_task` definition.  Therefore, this should work:

```python
new_task = [io_item if j == i else v for j, v in enumerate(dsk[k])]
dsk[k] = tuple(new_task)
```

Reply:
Yes, that does the trick